schemas/metadata-schema.json

```diff
@@ -0,0 +1,402 @@
+      "oneOf": [
+        {
+          "type": "number"
+        },
+        {
+          "type": "string"
+        }
+      ]
+    },
+    "storage": {
+      "description": "charm storage",
+      "type": "object",
+      "required": [ "type" ],
+      "additionalProperties": false,
+      "properties": {
+        "type": {
+          "$ref": "#/definitions/storageType"
+        },
+        "description": {
+          "type": "string",
+          "description": "Description of the storage requested"
+        },
+        
+        "shared": {
+          "type": "boolean",
+          "description": "True indicates all units of the application share the storage"
+        },
+        
+        "read-only": {
+          "type": "boolean",
+          "description": "True indicates the storage should be made read-only where possible"
+        },
+        
+        "multiple": {
+          "$ref": "#/definitions/storageCount"
+        },
+        
+        "minimum-size": {
+          "type": "string",
+          "description": "Size in the forms 1.0G, 1GiB, 1.0GB. Size multipliers are M, G, T, P, E, Z or Y. If no multiplier is supplied, M is implied."
+        },
+        
+        "location": {
+          "type": "string",
+          "description": "Location is the mount location for filesystem stores. For multi-stores, the location acts as the parent directory for each mounted store."
+        },
+        
+        "properties": {
+          "type": "array",
+          "description": "List of properties. Currently only supported value is 'transient'.",
+          "items": {
+            "type": "string",
+            "enum": [ "transient" ]
+          }
+        }
+        
+      }
+    },
+    "device": {
+      "type": "object",
+      "description": "Device requests for the charm, e.g. GPU. Key represents the Device name.",
+      "additionalProperties": {
+        "type": "object",
+        "required": [ "type" ],
+        "properties": {
+          "type": {
+            "type": "string",
+            "description": "Type of device requested. Currently supported values: 'gpu', 'nvidia.com/gpu' or 'amd.com/gpu'",
+            "enum": [ "gpu", "nvidia.com/gpu", "amd.com/gpu" ]
+          },
+          "description": {
+            "type": "string",
+            "description": "Description of the device requested."
+          },
+          "countmin": {
+            "type": "number",
+            "description": "Minimum number of devices required"
+          },
+          "countmax": {
+            "type": "number",
+            "description": "Maximum number of devices required"
+          }
+        }
+      }
+    },
+    "base": {
+      "type": "object",
+      "description": "A list of bases in descending order of preference for use in resolving a container image. Must not be present if resource is specified. These bases are listed as base (instead of name) and channel as in the Base definition, as an unnamed top-level object list.",
```

Comment:
```suggestion
      "description": "List of bases in descending order of preference for use in resolving a container image. Must not be present if resource is specified. These bases are listed as base (instead of name) and channel as in the Base definition, as an unnamed top-level object list.",
```